src/sqlparser.rs

```diff
@@ -448,6 +436,100 @@ impl Parser {
         })
     }
 
+    pub fn parse_date_time_field(&mut self) -> Result<SQLDateTimeField, ParserError> {
+        let tok = self.next_token();
+        if let Some(Token::SQLWord(ref k)) = tok {
+            match k.keyword.as_ref() {
+                "YEAR" => Ok(SQLDateTimeField::Year),
+                "MONTH" => Ok(SQLDateTimeField::Month),
+                "DAY" => Ok(SQLDateTimeField::Day),
+                "HOUR" => Ok(SQLDateTimeField::Hour),
+                "MINUTE" => Ok(SQLDateTimeField::Minute),
+                "SECOND" => Ok(SQLDateTimeField::Second),
+                _ => self.expected("date/time field", tok)?,
+            }
+        } else {
+            self.expected("date/time field", tok)?
+        }
+    }
+
+    /// Parse an INTERVAL literal.
+    ///
+    /// Some valid intervals:
+    ///     1. `INTERVAL '1' DAY`
+    ///     2. `INTERVAL '1-1' YEAR TO MONTH`
+    ///     3. `INTERVAL '1' SECONDS`
+    ///     4. `INTERVAL '1:1:1.1' HOUR (5) TO SECONDS (5)`
+    ///     5. `INTERVAL '1.1` SECONDS (2, 2)`
+    ///     6. `INTERVAL '1:1' HOUR (5) TO MINUTE (5)`
+    ///     7. `INTERVAL '1:1' SECOND TO SECOND`
+    ///
+    /// Note that (6) is not technically standards compliant, as the only
+    /// end qualifier which can specify a precision is `SECOND`. (7) is also
+    /// not standards compliant, as `SECOND` is not permitted to appear as a
+    /// start qualifier, except in the special form of (5). In the interest of
+    /// sanity, for the time being, we accept all the forms listed above.
+    pub fn parse_literal_interval(&mut self) -> Result<ASTNode, ParserError> {
+        // The first token in an interval is a string literal which specifies
+        // the duration of the interval.
+        let value = self.parse_literal_string()?;
+
+        // Following the string literal is a qualifier which indicates the units
+        // of the duration specified in the string literal.
+        let start_field = self.parse_date_time_field()?;
+
+        // The start qualifier is optionally followed by a numeric precision.
```

Comment:
> ...unless we need to support SECOND TO SECOND.

Nope, not for any reason I can think of! I think it's actually better to reject it, since I can't find a database that supports it, and the spec certainly seems to think it shouldn't be supported.

Reply:
I simplified this using your approach, but added a bit of complexity to not parse the fractional_seconds_precision unless the last field was actually seconds.